### PR TITLE
Shadows

### PR DESCRIPTION
**Pull request by SJuliez**
- Changes to how the hexes are drawn so that there are fewer pixel gaps
- Shadowmap is now a ton faster
- Dusk/Moonlit Night/Moonless Night use different shadow/overlay settings that make them look distinct. (no user control yet)

### CODE REVIEW

**Comment by arlith** [comment]
Looks like you're missing pitch black.  Just full moon and moonless are dark.  Dusk/dawn could also use a light level, since it's easier to make adjustments now.

**Comment by Akjosch** [comment]
There's a bunch of variable names in here (and in previous commits) which don't follow the style guide, by the way - the variables `DownArrow`,  `LightDirection`, `ShadowMap` and `UpArrow` in `BoardView1.java`  should be named  `downArrow`,  `lightDirection`, `shadowMap` and `upArrow`, respectively. I didn't check the other files.

**Comment by SJuliez** [comment]
style guide corrections added. They seem to add to the pull request so I wont do anything else.

**Comment by arlith** [comment]
I do still want to see something address for bitch black.  Right now it doesn't do any shading.

It might be nice to have some shading done for dawn/dusk as well, but I don't know that it is necessary.

(Your previous comment made it seem like you might not be adding anymore to the pull request; maybe that wasn't the intent)

**Comment by SJuliez** [comment]
bitch black - any day ^^


but seriously I am. I'll push it when ready.




________________________________
Von: Nicholas Walczak <notifications@github.com>
Gesendet: Mittwoch, 2. März 2016 11:03
An: MegaMek/megamek
Cc: SJuliez
Betreff: Re: [megamek] Shadows (#50)


I do still want to see something address for bitch black. Right now it doesn't do any shading.

It might be nice to have some shading done for dawn/dusk as well, but I don't know that it is necessary.

(Your previous comment made it seem like you might not be adding anymore to the pull request; maybe that wasn't the intent)

-

**Comment by SJuliez** [comment]
Ideas: (btw the shadow map is not used for this currently because it isn't applied over building roofs and woods or else these would mostly be in shadow which would most of the time be wrong)

The night things are pretty extreme so they'd need an easy toggle to switch off. (but note they're not quite as extreme as they look in screenshots with a white browser background around them)

Dawn: (yellowish light and longer shadows)

dawn

Pitch Black: (desaturated)

pitchblack

or.... with grain
pitchblack_grained

Full Moon: (bluish with blurry shadowing)

fullmoon

**Comment by pheonixstorm** [comment]
Looks good but night shots.. While I do like them they seem (as noted) too dark. Might have to build a copy and play on it to see how it looks in game.

**Comment by SJuliez** [comment]
@pheonixstorm I'll put up some full-size screenshots so you can look at them more easily. But not today (bedtime).

**Comment by arlith** [comment]
You might also want to test with having some searchlights and/or flares (basically, something to illuminate the hexes, to see how the illumination works with the night effects).

I think the full moon effect looks really cool.  This is completely gravy, but you could also potentially look at fog.  Rain/snow would be cool as well, but I think without animation they won't look quite right (and animation is not a great route to go, unless you know of a better way to do it then what I had explored in the past).

**Comment by HammerGS** [comment]
As Arlith mention, I'd love to see some type of weather affects on the maps Fog/Rain/Snow.

I also really like the look of the dawn, like the Pitch Black: (desaturated) and Full Moon: (bluish with blurry shadowing).

**Comment by SJuliez** [comment]
Fog is on the todo list. I believe rain/snow as Arlith said will not work without animation and I think would get pretty tedious very soon with animation. However Battle Isle 2 used to do a pretty good job (considering the hexes were like 16x16 pixels at 16 colors or so) giving the impression of water puddles and snow heaps. Maybe it's possible to do something like that. Not certain I'm artist enough for it...

**Comment by pheonixstorm** [comment]
Another point, since this isn't a real game engine adding 2D or 3D animation and effects may be pushing it too far. Would be nice, but... well, how fast does java run when dealing with OpenGL? I think using some type of particle engine (or manager) would give us some very nice smoke/fog rain/snow etc and we don't need an artist for smoke or fog so long as you don't mind on the fly texture rendering. Its more than possible in C++ so I think the only issue would be if there is a performance hit in Java.

**Comment by Akjosch** [comment]
@pheonixstorm OpenGL works rather well with Java, as Minecraft (which uses lwjgl) shows.I tried to see how well that would work with `IBoardView` already, and I remember there were a few cases where the interface was not so ideal for driving an OpenGL display, but I did get it as far as displaying some test 3D scene.

Also, there are cases where classes outside `megamek.client.ui.swing` use classes in that package, which might be a problem for a drop-in replacements too (`megamek.MegaMek` itself uses `megamek.client.ui.swing.ButtonOrderPreferences`, for example)

**Comment by arlith** [comment]
There was some refactoring I had wanted to do a long time ago to facilitate creating drop-in UI replacements... One thing that needs to happen is the command that define what buttons appear for each phase display need to get refactored ou of their respective phase displays and into non client.ui.swing classes.

**Comment by arlith** [comment]
To help me keep track of things, I added a new "pending" label, which I'm using to indicate PRs that are still being worked on.

**Comment by SJuliez** [comment]
@HammerGS Ok, the shadowmap() is the culprit for slowness. I thought I had the obvious check in if the shadowmap is settings-activated at all but I hadnt :flushed: 
That is easily corrected (see pull request), but the TODO remains to make it faster and to have it not readraw unless absolutely necessary (=when the board changes).

**Comment by SJuliez** [comment]
Updated. Much faster at a little less quality. Will now also not redraw the shadowmap when all units have been moved but only when a hex is changed, i.e. not so often. If redrawn will redraw in a few ms on smaller maps, <1s on big maps, ~2s or so on really big maps with 1000s of woods or crazy elevations. (Devils Tower: 2.2s)

**Comment by arlith** [comment]
So you're ready to have this PR reviewed and merged then, yes?

**Comment by SJuliez** [comment]
Let me test this by playing half a game on some interesting map. Havent done that yet.

**Comment by Akjosch** [comment]
Here's one "interesting" map for you. We played that one once, together with my "ground map elevation translates to aerospace altitude" patch to make things a bit varied for the aero jockeys. 160x160 hexes. Almost 100 height altitude difference.

[crazycanyon.zip](https://github.com/MegaMek/megamek/files/168158/crazycanyon.zip)

**Comment by SJuliez** [comment]
@Akjosch your maps give me a headache :smile: I think the first time I called up the crazycanyon I killed the MM process after 2 minutes. OK I went through 4 or 5 different ways to draw the elevation shadows but now I have it down to 3 or 4s (only the shadowmap; drawing the hexes takes another 2s or so, so all in all it does take maybe 8s until its all ready). I took the liberty of limiting the level difference that the shadows are drawn with to 35. Helps with speed, protects against freezes when someone uses a map with a level 2100 hexes and we don't want the whole map in shadow anyway.

Your other map has a huge load of woods. It also takes like 5s to draw the shadows for that map, I think Im going to work on optimizing that.
@arlith If you don't mind, please do keep it pending for now.

**Comment by Akjosch** [comment]
@SJuliez And the best thing is, those are the maps we actually *used*, not something done specifically to mess with you. ;)

**Comment by arlith** [comment]
I think that there are some fixes in here that should probably go into MM if there is a new release, so I might merge this branch, and you can add more later.

**Comment by Akjosch** [comment]
From my side, it would also be nice to know if you plan to do anything with TilesetManager or MechTileset. I have some plans to make them both load parts of an image (format: "image.png[123x456]" or "image.png[123x456:84x72]" if you want specific size differing from 84x72), allowing to create texture atlases instead of millions of tiny image files all over the place, but that's a bigger change in how images are loaded and stored, so I might easily get into conflicts with whatever it is you're doing there.

**Comment by arlith** [comment]
If I understand you correctly, a patch Juliez submitted a while back allows for this already.  I think there's a largeTextures tileset or something like that that illustrates this.  It allows you to specify a larger texture, and the hex can figure out where it is in that larger texture.

See this post here: http://megamek.info/forums/index.php?topic=1994.msg12954#msg12954

**Comment by Akjosch** [comment]
Not quite, what I plan to do is create a patch with the following changes:

* You can put multiple hex images into one image (tileset and unit files) and have the configuration tell MM which part of the image to cut out for a specific hex' or unit's image
* Generic resource manager to never load the same image more than once, no matter how many units use it.
* Allow to write simple plugins for loading image types (and naming schemas) which `Toolkit.getImage(fileName)` doesn't support.

**Comment by SJuliez** [comment]
@arlith Alright, we'll see if any problems come up and the shadow system is certainly better than it used to be.

@Akjosch OK, I can see this reducing the 20000 or so files that MM comes with :) It might however need some kind of util to pack the existing files and generate the config for it. A question though: isnt MM able to work with ZIP files anyway? (At least in some corner of it) That would solve the problem of file numbers and no config would be necessary.

Other things that crossed my mind concerning tiles:
- I would actually like to implement a system to handle terrain supers that exceed the limits of the hex tile (a little). Trees are quite limited by the current system. That's a bit incompatible with the hex cache though. Two ways: a) redraw the hex cache to an additional buffer that will then be painted in paintComponent. Anything can be painted into that. No performance hit, but more memory necessary, more repaint management and I dont know how that interferes with isometric. b) convert anything bigger than 84x72 to a sprite. Not good for performance but I have repaint times of 9ms so even a few hundred sprites over that wont be noticable. Your woods-y savegame would again be the benchmark.
- I was thinking about adding support for unit icons bigger than the usual 84x72. It would assume that anything bigger than 84x72 is to be compressed to the current hex size when painted. Another thought was to add support for svg unit icons. I tested converting unit icons to svg at some websites and the results are actually *far* better than I expected from those small icons (seriously). Converting 5000 of them would be some effort though and Im not sure it's doable for free (with some quality). So here the main question would be: is Deadborder happy with how icons are currently done or would he actually paint them in vector if possible or paint them bigger if useful? 
- I was at a time thinking about adding the ability for MM to generate road tiles by superposition and rotation of basic road tiles. It takes just 3 different images (straight, long curve, short curve) to generate all unbroken street types and crossings. The quality of crossings is reduced vs. hand-painted 64 images but it's faaaar more accessible. I have a POVRay script to render them and I can create loads of differently textured roads but for the crossings it uses exactly that system. On the other hand, I do have that script and it automatically generates the 64 images (minus the broken off street ones) so that's not an important feature.

**Comment by Akjosch** [comment]
Actually, it takes *four* hex images to be able to create the other 60 out of them for a full set, the "broken" ends included: One "no connection" hex, one "fully connected" hex, one "Y" hex and one "inverted Y" hex.

hexdisasembly

You can easily pack all images into zip files, but you will still need to do something different than `Toolkit.getImage(fileName)`, which is where the plugins to load images from different naming schemas come in.

**Comment by SJuliez** [comment]
Interesting image that reminds me that something I rather want to do is create a system to smooth out interfaces between different base textures a bit, e.g. shorelines, rivers, water depths. Won't mess with the tilesets tho.

**Comment by HammerGS** [comment]
I think anything that improves the look and feel of MegaMek is a good thing.  I know Arlith and I have had numerous conversations about it.

Personally, I use ISO a lot would would love to some more variety in terms of:

Hills - removing the grey blocking and making them look more hill like (more rounded)

Buildings - Have them look more like buildings again getting rid of the grayness

Water - Have something that reflects the depth of the water

Basically getting rid of the gray.

I'll drop a note to Deadboarder about:

_"So here the main question would be: is Deadborder happy with how icons are currently done or would he actually paint them in vector if possible or paint them bigger if useful?"_

and will post his reply.

**Comment by arlith** [comment]
Another thing to keep in mind is that a 5-9ms redraw time with an i7 isn't going to be the same with older hardware.  I think a fair amount of MM users aren't playing on new fancy hardware, so we should keep this in mind for graphics performance.

**Comment by Deadborder** [comment]
I'll admit that I'm happy with how Icons are done at the moment for two main reasons.

The first is that I have my process rather down-pat and functional, to the point where I can ripple off a new icon (or group for a 'Mech with multiple variants) in about 30-45 minutes depending on size and complexity. It's pretty easy and something I can breezily do at the end of my working day or whenever else I find a bit of time. And it produces nicely functional results.

Learning a new process is probably going to take a lot of time, and slow down the rate of production (as well as probably cause a quality drop until I get back on top of it).

The second reason is simple economics of scale. I've drawn ~2000 icons so far, and re-drawing/converting/re-scaling them is going to take a lot of time. I personally think it'd be better to keep the current process and focus on other areas of MM development for now.

**Comment by Akjosch** [comment]
That's the general issue with "indie" game development: The availability of quality artwork (and content in general, like voice acting, music, story, dialogue) is very much a limiting factor, not the amount of ideas or people coding. So we can't be thankful enough to @Deadborder for the mountain of work he did so far.

Sadly, that means any optional icon standard will remain underdeveloped for a *long* time. The more complex the worse; I'd be happy if we get just a few dozen basic low-poly 3D mechs, for example.

**Comment by beerockxs** [comment]
I'll take a look on my old laptop.

Am 13. März 2016 22:25:00 MEZ, schrieb Nicholas Walczak <notifications@github.com>:
>Another thing to keep in mind is that a 5-9ms redraw time with an i7
>isn't going to be the same with older hardware.  I think a fair amount
>of MM users aren't playing on new fancy hardware, so we should keep
>this in mind for graphics performance.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/MegaMek/megamek/pull/50#issuecomment-196056009

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by SJuliez** [comment]
@Deadborder OK thanks! Redoing any of the present icons which are wonderful and one of the strong points of MM was for me out of the question (except if it was possible to automate it); this was only for future icons. Since you are happy with the present system there is no need to add any functionality to it for now.

**Comment by arlith** [comment]
@SJuliez 

What do you think about having different shadows based on illumination levels at night?  For instance, see this screenshot: http://i.imgur.com/7RJRLOM.jpg.  This came from the attached saved game (which came from #81).  One could imagine that the shadows are different for the forest in the illuminated areas.

[GrappleBug-SubmitLater.sav.gz](https://github.com/MegaMek/megamek/files/182872/GrappleBug-SubmitLater.sav.gz)

As an aside, you can enable the Issue tracker in your fork by going to Settings.  It could be useful for you to enable that, and I could post thoughts/ideas to the issue tracker, instead of tacking them on as comments to PRs.

**Comment by SJuliez** [comment]
OK Issue tracker is activated.

I did think about lighting when I tested flares but didnt do anything yet:
- the cast shadows (that arent very obvious in the screenshot, either deactivated or moonless night I guess) should probably be deactivated within the illuminated area so it doesnt look as if the sun came out -> one for the TODO list.
- at first glance I would say that changing the AO shadows visible in the screenshot would look confusing; Id rather not do that
- it might be possible to create some sort of a flare centered shadow in its own small buffer; I'd expect (but I dont know) a flare to be a few levels above the ground and create strongly contrasted shadows. The light direction depends on the hex relative to the flare and the shadows expand with distance -> one for the wish list.
